### PR TITLE
fuzz: More accurate coverage reports

### PR DESCRIPTION
**Pull request by dergoegge**
Our coverage reports include coverage of initialization code, which can be misleading when trying to evaluate the coverage a fuzz harness achieves through fuzzing alone.

This PR proposes to make fuzz coverage reports more accurate by resetting coverage counters after initialization code has been run. This makes it easier to evaluate which code was actually reached through fuzzing (e.g. to spot fuzz blockers).

### CODE REVIEW

**Comment by TheCharlatan** [comment]
Concept ACK

File: src/test/fuzz/fuzz.cpp [hunk 1]
```diff
@@ -185,13 +199,20 @@ extern "C" int LLVMFuzzerInitialize(int* argc, char*** argv)
...
     return 0;
 }
 
+
 #if defined(PROVIDE_FUZZ_MAIN_FUNCTION)
 int main(int argc, char** argv)
 {
     initialize();
+
+    ResetCoverageCounters();
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/test/fuzz/fuzz.cpp]: No reason, done!

**Comment by hebasto** [comment]
Concept ACK on improving coverage.

**Comment by kevkevinpal** [comment]
Concept ACK [52506a0](https://github.com/bitcoin/bitcoin/pull/30156/commits/52506a0bb28a0c052deef9f484dc94e83de4f0f7)

Made a clean build of this change using
```
sudo make clean && ./autogen.sh &&  CC=clang CXX=clang++ ./configure --enable-fuzz --with-sanitizers=address,fuzzer,undefined && sudo make -j"$(($(nproc)+1))" install
```

## Running fuzz tests twice

(master: f15778536ad421f9805f0c005f0eb7b84dda1b4e)
```
FUZZ=process_message src/test/fuzz/fuzz -runs=1 qa-assets/fuzz_seed_corpus/process_message
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 4218378445
INFO: Loaded 1 modules   (425045 inline 8-bit counters): 425045 [0x558dbe652ce0, 0x558dbe6ba935),
INFO: Loaded 1 PC tables (425045 PCs): 425045 [0x558dbe6ba938,0x558dbed36e88),
INFO:     2759 files found in qa-assets/fuzz_seed_corpus/process_message
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 814557 bytes
INFO: seed corpus: files: 2759 min: 1b max: 814557b total: 14744612b rss: 218Mb
#1024   pulse  cov: 11518 ft: 23753 corp: 583/80Kb exec/s: 341 rss: 454Mb
#2048   pulse  cov: 13725 ft: 35158 corp: 1025/223Kb exec/s: 256 rss: 454Mb
#3698   INITED cov: 14297 ft: 52528 corp: 1782/8638Kb exec/s: 160 rss: 581Mb
#3698   DONE   cov: 14297 ft: 52528 corp: 1782/8638Kb lim: 630378 exec/s: 160 rss: 581Mb
Done 3698 runs in 23 second(s)

FUZZ=process_message src/test/fuzz/fuzz -runs=1 qa-assets/fuzz_seed_corpus/process_message
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 555650094
INFO: Loaded 1 modules   (425045 inline 8-bit counters): 425045 [0x557c8bfd0ce0, 0x557c8c038935),
INFO: Loaded 1 PC tables (425045 PCs): 425045 [0x557c8c038938,0x557c8c6b4e88),
INFO:     2759 files found in qa-assets/fuzz_seed_corpus/process_message
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 814557 bytes
INFO: seed corpus: files: 2759 min: 1b max: 814557b total: 14744612b rss: 219Mb
#3696   INITED cov: 14264 ft: 52459 corp: 1784/8568Kb exec/s: 160 rss: 583Mb
#3696   DONE   cov: 14264 ft: 52459 corp: 1784/8568Kb lim: 630378 exec/s: 160 rss: 583Mb
Done 3696 runs in 23 second(s)
```

(dergoegge:2024-05-cov-reset-counters: [52506a0](https://github.com/bitcoin/bitcoin/pull/30156/commits/52506a0bb28a0c052deef9f484dc94e83de4f0f7))
```
FUZZ=process_message src/test/fuzz/fuzz -runs=1 qa-assets/fuzz_seed_corpus/process_message
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1724170080
INFO: Loaded 1 modules   (424990 inline 8-bit counters): 424990 [0x55773dc1cc20, 0x55773dc8483e),
INFO: Loaded 1 PC tables (424990 PCs): 424990 [0x55773dc84840,0x55773e300a20),
INFO:     2759 files found in qa-assets/fuzz_seed_corpus/process_message
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 814557 bytes
INFO: seed corpus: files: 2759 min: 1b max: 814557b total: 14744612b rss: 219Mb
#1024   pulse  cov: 11499 ft: 23802 corp: 589/81Kb exec/s: 341 rss: 455Mb
#2048   pulse  cov: 13711 ft: 35251 corp: 1034/226Kb exec/s: 256 rss: 455Mb
#3701   INITED cov: 14277 ft: 52623 corp: 1793/8700Kb exec/s: 160 rss: 582Mb
#3701   DONE   cov: 14277 ft: 52623 corp: 1793/8700Kb lim: 630378 exec/s: 160 rss: 582Mb
Done 3701 runs in 23 second(s)

FUZZ=process_message src/test/fuzz/fuzz -runs=1 qa-assets/fuzz_seed_corpus/process_message
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 3858639904
INFO: Loaded 1 modules   (424990 inline 8-bit counters): 424990 [0x562169278c20, 0x5621692e083e),
INFO: Loaded 1 PC tables (424990 PCs): 424990 [0x5621692e0840,0x56216995ca20),
INFO:     2759 files found in qa-assets/fuzz_seed_corpus/process_message
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 814557 bytes
INFO: seed corpus: files: 2759 min: 1b max: 814557b total: 14744612b rss: 219Mb
#1024   pulse  cov: 11480 ft: 23795 corp: 578/80Kb exec/s: 341 rss: 456Mb
#2048   pulse  cov: 13698 ft: 35272 corp: 1017/222Kb exec/s: 256 rss: 456Mb
#3700   INITED cov: 14266 ft: 52596 corp: 1769/8685Kb exec/s: 160 rss: 582Mb
#3700   DONE   cov: 14266 ft: 52596 corp: 1769/8685Kb lim: 630378 exec/s: 160 rss: 582Mb
Done 3700 runs in 23 second(s)
```

**Comment by instagibbs** [comment]
concept ACK

File: src/test/fuzz/fuzz.cpp [hunk 2]
```diff
@@ -79,6 +79,26 @@ void FuzzFrameworkRegisterTarget(std::string_view name, TypeTestOneInput target,
 static std::string_view g_fuzz_target;
 static const TypeTestOneInput* g_test_one_input{nullptr};
 
+
+#if defined(__clang__) && defined(__linux__)
```
Comment (dergoegge) [inline src/test/fuzz/fuzz.cpp:83]: I don't know how to make this work outside of these params. I don't want to use a macro, I think this should always be enabled if possible.

Comment (dergoegge) [reply src/test/fuzz/fuzz.cpp:83]: https://cirrus-ci.com/task/6736987506343936:

```
Undefined symbols for architecture x86_64:
  "___llvm_profile_reset_counters", referenced from:
      initialize() in libtest_fuzz.a(libtest_fuzz_a-fuzz.o)
  "___gcov_reset", referenced from:
      initialize() in libtest_fuzz.a(libtest_fuzz_a-fuzz.o)
ld: symbol(s) not found for architecture x86_64
```

https://cirrus-ci.com/task/6596250017988608:

```
/usr/bin/ld: libtest_fuzz.a(libtest_fuzz_a-fuzz.o): in function `ResetCoverageCounters()':
./src/test/fuzz/fuzz.cpp:90: undefined reference to `__llvm_profile_reset_counters'
/usr/bin/ld: ./src/test/fuzz/fuzz.cpp:94: undefined reference to `__gcov_reset'
/usr/bin/ld: ./src/test/fuzz/fuzz.cpp:90: undefined reference to `__llvm_profile_reset_counters'
/usr/bin/ld: ./src/test/fuzz/fuzz.cpp:94: undefined reference to `__gcov_reset'
collect2: error: ld returned 1 exit status
```

https://cirrus-ci.com/task/5470350111145984:

```
test/fuzz/fuzz.cpp:84:1: error: attributes are not permitted in this position [-Werror=attributes]
   84 | __attribute__((weak)) extern "C" void __llvm_profile_reset_counters(void);
      | ^~~~~~~~~~~~~
test/fuzz/fuzz.cpp:84:33: note: attributes may be inserted here
   84 | __attribute__((weak)) extern "C" void __llvm_profile_reset_counters(void);
      |                                 ^
test/fuzz/fuzz.cpp:85:1: error: attributes are not permitted in this position [-Werror=attributes]
   85 | __attribute__((weak)) extern "C" void __gcov_reset(void);
      | ^~~~~~~~~~~~~
test/fuzz/fuzz.cpp:85:33: note: attributes may be inserted here
   85 | __attribute__((weak)) extern "C" void __gcov_reset(void);
      |                                 ^
test/fuzz/fuzz.cpp: In function ‘void ResetCoverageCounters()’:
test/fuzz/fuzz.cpp:89:9: error: the address of ‘void __llvm_profile_reset_counters()’ will never be NULL [-Werror=address]
   89 |     if (__llvm_profile_reset_counters) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test/fuzz/fuzz.cpp:84:39: note: ‘void __llvm_profile_reset_counters()’ declared here
   84 | __attribute__((weak)) extern "C" void __llvm_profile_reset_counters(void);
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test/fuzz/fuzz.cpp:93:9: error: the address of ‘void __gcov_reset()’ will never be NULL [-Werror=address]
   93 |     if (__gcov_reset) {
      |         ^~~~~~~~~~~~
test/fuzz/fuzz.cpp:85:39: note: ‘void __gcov_reset()’ declared here
   85 | __attribute__((weak)) extern "C" void __gcov_reset(void);
      |                                       ^~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

**Comment by dergoegge** [comment]
As an example for the minisketch harness, files reported as reached by the fuzzer:

master:

```
src/coins.h
src/common/args.cpp
src/common/system.cpp
src/compat/byteswap.h
src/compat/endian.h
src/crypto/chacha20.cpp
src/crypto/chacha20.h
src/crypto/common.h
src/crypto/sha256.cpp
src/crypto/sha512.cpp
src/crypto/sha512.h
src/crypto/siphash.cpp
src/cuckoocache.h
src/hash.cpp
src/hash.h
src/logging.cpp
src/logging.h
src/minisketch/include/minisketch.h
src/minisketch/src/fields/generic_4bytes.cpp
src/minisketch/src/fields/generic_common_impl.h
src/minisketch/src/int_utils.h
src/minisketch/src/lintrans.h
src/minisketch/src/minisketch.cpp
src/minisketch/src/sketch.h
src/minisketch/src/sketch_impl.h
src/net.cpp
src/netaddress.cpp
src/netbase.h
src/policy/feerate.h
src/prevector.h
src/primitives/transaction.h
src/pubkey.cpp
src/random.cpp
src/randomenv.cpp
src/rpc/client.cpp
src/rpc/server.cpp
src/rpc/server.h
src/rpc/util.cpp
src/rpc/util.h
src/script/miniscript.cpp
src/script/miniscript.h
src/script/script.h
src/script/sigcache.cpp
src/script/sign.cpp
src/script/solver.cpp
src/secp256k1/src/hash_impl.h
src/secp256k1/src/secp256k1.c
src/secp256k1/src/selftest.h
src/secp256k1/src/util.h
src/serialize.h
src/span.h
src/support/allocators/secure.h
src/support/cleanse.cpp
src/support/lockedpool.cpp
src/support/lockedpool.h
src/sync.h
src/test/fuzz/FuzzedDataProvider.h
src/test/fuzz/fuzz.cpp
src/test/fuzz/fuzz.h
src/test/fuzz/minisketch.cpp
src/test/fuzz/script_assets_test_minimizer.cpp
src/test/fuzz/txrequest.cpp
src/test/fuzz/util.h
src/test/util/script.h
src/test/util/setup_common.cpp
src/uint256.h
src/univalue/include/univalue.h
src/univalue/lib/univalue.cpp
src/univalue/lib/univalue_read.cpp
src/util/check.h
src/util/fs.cpp
src/util/fs.h
src/util/spanparsing.h
src/util/strencodings.cpp
src/util/strencodings.h
src/util/string.h
src/util/threadinterrupt.cpp
src/util/time.cpp
src/util/time.h
src/util/vector.h
```

pull:

```
src/minisketch/include/minisketch.h
src/minisketch/src/fields/generic_4bytes.cpp
src/minisketch/src/fields/generic_common_impl.h
src/minisketch/src/int_utils.h
src/minisketch/src/lintrans.h
src/minisketch/src/minisketch.cpp
src/minisketch/src/sketch.h
src/minisketch/src/sketch_impl.h
src/span.h
src/test/fuzz/FuzzedDataProvider.h
src/test/fuzz/fuzz.cpp
src/test/fuzz/minisketch.cpp
src/test/fuzz/util.h
src/util/check.h
src/util/fs.cpp
src/util/fs.h
src/util/time.h
```

**Review by brunoerg — APPROVED** [review]
nice, utACK 949abebea0059edd929b653b4b475a5880fc0a3e

**Comment by dergoegge** [comment]
rfm?

File: src/test/fuzz/fuzz.cpp (hunk 2 above)
Comment (fanquake) [reply src/test/fuzz/fuzz.cpp:83]: Might followup here.

Comment (hodlinator) [reply src/test/fuzz/fuzz.cpp:83]: (Worked on in #32059).

**Pull request merged**